sc-poc/managers/configman/adapter/k8s/k8s.go

```diff
@@ -0,0 +1,213 @@
+		factory.ForResource(graphqlsourcesResource).Informer(),
+		factory.ForResource(jwthsasecretsResource).Informer(),
+		factory.ForResource(opapoliciesResource).Informer())
+
+	k := &K8s{
+		dc:             dc,
+		namespace:      namespace,
+		logger:         logger,
+		informers:      informers,
+		configuration:  make(map[string][]*unstructured.Unstructured),
+		configurationN: make(map[string][]*unstructured.Unstructured),
+	}
+
+	return k, nil
+}
+
+func (k *K8s) GetRawConfig() ([]byte, error) {
+	// Load SC config file from cluster
+	err := k.loadConfiguration()
+	if err != nil {
+		return nil, err
+	}
+
+	// Load the new caddy config
+	return k.toRaw()
+}
+
+func (k *K8s) Run(ctx context.Context) (chan []byte, error) {
+	cfgChan := make(chan []byte)
+
+	for _, informer := range k.informers {
+		informer.AddEventHandler(k8sCache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				u := obj.(*unstructured.Unstructured)
+
+				// new configuration
+				key := source.GetModuleName(u.GetAPIVersion(), u.GetKind())
+				found := false
+				for i, spec := range k.configurationN[key] {
+					if spec.GetName() == u.GetName() {
+						k.configurationN[key][i] = u
+						found = true
+						break
+					}
+				}
+				if !found {
+					k.configurationN[key] = append(k.configurationN[key], u)
+				}
+
+				// old configuration
+				kind := u.GetKind()
+				found = false
+				for i, spec := range k.configuration[kind] {
+					if spec.GetName() == u.GetName() {
+						k.configuration[kind][i] = u
+						found = true
+						break
+					}
+				}
+				if !found {
+					k.configuration[kind] = append(k.configuration[kind], u)
+				}
+
+				resp, err := k.toRaw()
+				if err != nil {
+					k.logger.Error("error reloading config: ", zap.Error(err))
+					return
+				}
+
+				cfgChan <- resp
+			},
+			UpdateFunc: func(_ interface{}, newObj interface{}) {
+				u := newObj.(*unstructured.Unstructured)
+
+				// new configuration
+				key := source.GetModuleName(u.GetAPIVersion(), u.GetKind())
+				for i, spec := range k.configurationN[key] {
+					if spec.GetName() == u.GetName() {
+						k.configurationN[key][i] = u
+						break
+					}
+				}
+
+				// old configuration
+				kind := u.GetKind()
+				for i, spec := range k.configuration[kind] {
+					if spec.GetName() == u.GetName() {
+						k.configuration[kind][i] = u
+						break
+					}
+				}
+
+				resp, err := k.toRaw()
+				if err != nil {
+					k.logger.Error("error reloading config: ", zap.Error(err))
+					return
+				}
+
+				cfgChan <- resp
+			},
+			DeleteFunc: func(obj interface{}) {
+				u := obj.(*unstructured.Unstructured)
+
+				// new configuration
+				key := source.GetModuleName(u.GetAPIVersion(), u.GetKind())
+				s := []*unstructured.Unstructured{}
+				for _, spec := range k.configurationN[key] {
+					if spec.GetName() == u.GetName() {
+						continue
+					}
+					s = append(s, spec)
+				}
+				k.configurationN[key] = s
```

Comment:
I suggest you use this method to remove to minimize allocs 
```golang
func remove(s []int, i int) []int {
    s[i] = s[len(s)-1]
    return s[:len(s)-1]
}
```

Reply:
this method doesn't work when length is 1